PlayCover/AppInstaller/Downloader.swift

```diff
@@ -0,0 +1,121 @@
+
+    func start() {
+        if downloadVM.downloading && installVM.installing {
+            Log.shared.error(PlayCoverError.waitDownload)
+        } else {
+            if let warningMessage = warning {
+                let alert = NSAlert()
+                alert.messageText = NSLocalizedString(warningMessage, comment: "")
+                alert.informativeText = String(
+                    format: NSLocalizedString("ipaLibrary.alert.download", comment: ""),
+                    arguments: [app.name]
+                )
+                alert.alertStyle = .warning
+                alert.addButton(withTitle: NSLocalizedString("button.Yes", comment: ""))
+                alert.addButton(withTitle: NSLocalizedString("button.No", comment: ""))
+
+                if alert.runModal() == .alertSecondButtonReturn {
+                    return
+                } else {
+                    proceedDownload()
+                }
+            } else {
+                proceedDownload()
+            }
+        }
+    }
+
+    func cancel() {
+        downloader.cancelAllDownloads()
+        downloadVM.downloading = false
+        downloadVM.progress = 0
+        downloadVM.storeAppData = nil
+    }
+
+    private func proceedDownload() {
+        var tmpDir: URL?
+        do {
+            tmpDir = try FileManager.default.url(for: .itemReplacementDirectory,
+                                                 in: .userDomainMask,
+                                                 appropriateFor: URL(fileURLWithPath: "/Users"),
+                                                 create: true)
+            downloader.addDownload(url: url,
+                                   destinationURL: tmpDir!,
+                                   onProgress: { progress in
+                self.downloadVM.storeAppData = self.app
+                self.downloadVM.downloading = true
+                // progress is a Float
+                self.downloadVM.progress = Double(progress)
+            }, onCompletion: { error, fileURL in
+                guard error == nil else {
+                    self.downloadVM.downloading = false
+                    self.downloadVM.progress = 0
+                    self.downloadVM.storeAppData = nil
+                    return Log.shared.error(error!)
+                }
+                self.downloadVM.downloading = false
+                self.downloadVM.progress = 0
+                self.proceedInstall(fileURL)
+            })
+        } catch {
+            if let tmpDir = tmpDir {
+                FileManager.default.delete(at: tmpDir)
+            }
+            Log.shared.error(error)
+        }
+    }
+
+    private func proceedInstall(_ url: URL?) {
+        if let url = url {
+            uif.ipaUrl = url
+            Installer.install(ipaUrl: uif.ipaUrl!, export: false, returnCompletion: { _ in
+                DispatchQueue.main.async {
+                    FileManager.default.delete(at: url)
+                    AppsVM.shared.apps = []
```

Comment:
Unnecessary, `fetchApps()` already clears the array

Reply:
It was in your own function, I suspect it would be necessary because if user press download and open AppLibrary after and wait for download there the list won't update. If it's not like that, tell me which line you want to be removed.